src/tsickle.ts

```diff
@@ -1360,27 +1374,26 @@ class Annotator extends ClosureRewriter {
   /**
    * @param optional If true, property is optional (e.g. written "foo?: string").
    */
-  private visitProperty(namespace: string[], prop: ts.Declaration, optional = false) {
+  private visitProperty(
+      namespace: string[], prop: ts.Declaration&{type?: ts.TypeNode}, optional = false) {
     const name = this.propertyName(prop);
     if (!name) {
       this.emit(`/* TODO: handle strange member:\n${this.escapeForComment(prop.getText())}\n*/\n`);
       return;
     }
 
-    let type = this.typeToClosure(prop);
+    let type = this.typeToClosure(prop, prop.type);
     // When a property is optional, e.g.
     //   foo?: string;
-    // Then the TypeScript type of the property is string|undefined, the
-    // typeToClosure translation handles it correctly, and string|undefined is
-    // how you write an optional property in Closure.
-    //
-    // But in the special case of an optional property with type any:
-    //   foo?: any;
-    // The TypeScript type of the property is just "any" (because any includes
-    // undefined as well) so our default translation of the type is just "?".
-    // To mark the property as optional in Closure it must have "|undefined",
+    // to mark the property as optional in Closure, it must have "|undefined",
     // so the Closure type must be ?|undefined.
-    if (optional && type === '?') type += '|undefined';
+    // The underlying TypeScript type of the property is e.g. "string|undefined"
+    // when a concrete type is given. However translation doesn't include
+    // "|undefined" in two corner cases:
+    // - for union types, we translate the syntactical type node
+    // - for any types, TS translation only includes ? (as any includes
+    //   "|undefined"), but Closure's ? type does not include absence
+    if (optional) type += '|undefined';
```

Comment:
This will add `|undefined` even in cases not explained above, unless there's some logic elsewhere I'm not seeing - maybe say why that's ok in the comment?a

Reply:
Added parentheses.